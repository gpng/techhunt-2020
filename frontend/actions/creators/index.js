export const APP_ACTIONS = {
  MODAL_CLOSE: 'modal_close',
  MODAL_OPEN: 'modal_open',
  UPLOAD_CSV: {
    SUBMIT: 'upload_submit',
    LOADING: 'upload_loading',
    LOADED: 'upload_loaded',
  },
  SEARCH: {
    SUBMIT: 'search_submit',
    LOADING: 'search_loading',
    LOADED: 'search_loaded',
  },
  DELETE: {
    SUBMIT: 'delete_submit',
    LOADING: 'delete_loading',
    LOADED: 'delete_loaded',
  },
  EDIT: {
    SUBMIT: 'edit_submit',
    LOADING: 'edit_loading',
    LOADED: 'edit_loaded',
  },
};

export const closeModal = () => ({ type: APP_ACTIONS.MODAL_CLOSE });

export const openModal = (name, value) => ({
  type: APP_ACTIONS.MODAL_OPEN,
  payload: { name, value },
});

export const submitUploadCSV = (file) => ({
  type: APP_ACTIONS.UPLOAD_CSV.SUBMIT,
  payload: file,
});

export const loadingUploadCSV = () => ({
  type: APP_ACTIONS.UPLOAD_CSV.LOADING,
});

export const loadedUploadCSV = (success, error) => ({
  type: APP_ACTIONS.UPLOAD_CSV.LOADED,
  payload: {
    success,
    error,
  },
});

export const submitSearch = (minSalary, maxSalary, sort, offset, limit) => ({
  type: APP_ACTIONS.SEARCH.SUBMIT,
  payload: { minSalary, maxSalary, sort, offset, limit },
});

export const loadingSearch = () => ({
  type: APP_ACTIONS.SEARCH.LOADING,
});

export const loadedSearch = (results, success, error) => ({
  type: APP_ACTIONS.SEARCH.LOADED,
  payload: {
    results,
    success,
    error,
  },
});

export const submitDelete = (id) => ({
  type: APP_ACTIONS.DELETE.SUBMIT,
  payload: id,
});

export const loadingDelete = () => ({
  type: APP_ACTIONS.DELETE.LOADING,
});

export const loadedDelete = (success, error) => ({
  type: APP_ACTIONS.DELETE.LOADED,
  payload: {
    success,
    error,
  },
});

export const submitEdit = (employee) => ({
  type: APP_ACTIONS.EDIT.SUBMIT,
  payload: employee,
});

export const loadingEdit = () => ({
  type: APP_ACTIONS.EDIT.LOADING,
});

export const loadedEdit = (success, error) => ({
  type: APP_ACTIONS.EDIT.LOADED,
  payload: {
    success,
    error,
  },
});
