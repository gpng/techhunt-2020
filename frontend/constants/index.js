export const API_HOST = 'http://localhost:4000';

export const BREAKPOINTS = {
  LARGE: 1024,
  MEDIUM: 768,
  SMALL: 600,
};

export const MODALS = {
  UPLOAD_CSV: 'upload_csv',
  DELETE: 'delete_employee',
};

export const SEARCH = {
  PAGE_SIZE: 5,
  DEFAULTS: {
    MIN_SALARY: 0,
    MAX_SALARY: 5000,
    PAGE: 0,
    SORT: '+id',
  },
};
