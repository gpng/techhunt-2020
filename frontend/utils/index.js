/**
 * Takes a promise, and returns then promised results in format [err, response]
 * @param { Promise } promise
 * @param { Object= } errorExt - Additional Information you can pass to the err object
 * @return { Promise }
 */
export const to = (promise, errorExt) => {
  return promise
    .then((data) => [null, data])
    .catch((err) => {
      if (errorExt) {
        Object.assign(err, errorExt);
      }

      return [err, undefined];
    });
};

export default to;
