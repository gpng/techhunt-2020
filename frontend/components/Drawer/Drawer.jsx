import React from 'react';
// components
// translations
import useTranslations from '../../translations/useTranslations';
import { COLORS } from '../../constants/styles';
import Avatar from '../Profile/Avatar';

const Drawer = () => {
  const { t } = useTranslations();

  return (
    <aside className="drawer-root">
      <Avatar width="6rem" />
      <div className="button-container">
        <button type="button" className="button-function">
          {t('drawer.functions.uploadCSV')}
        </button>
      </div>
      <style jsx>{`
        .drawer-root {
          background: ${COLORS.BACKGROUND_DARK_PRIMARY};
          height: 100%;
          width: 100%;
          display: flex;
          flex-direction: column;
          align-items: center;
          padding: 2rem 0;
        }

        .button-container {
          margin-top: 2rem;
          width: 100%;
        }

        .button-function {
          width: 100%;
          background: ${COLORS.BACKGROUND_DARK_SECONDARY};
          color: ${COLORS.TEXT_DARK_PRIMARY};
          border: none;
          padding: 0.5rem 0;
        }
      `}</style>
    </aside>
  );
};

export default Drawer;
