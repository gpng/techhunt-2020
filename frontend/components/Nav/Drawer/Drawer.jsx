import React, { useContext } from 'react';
// context
import AppContext from '../../Context/App/AppContext';
// components
import Avatar from '../../Profile/Avatar';
import LocaleDisplay from '../../Context/Locale/LocaleDisplay';
// translations
import useTranslations from '../../../translations/useTranslations';
// actions
import { openModal } from '../../../actions/creators';
// constants
import { MODALS } from '../../../constants';
import { COLORS } from '../../../constants/styles';

const Drawer = () => {
  const { t } = useTranslations();
  const { dispatch } = useContext(AppContext);

  return (
    <aside className="drawer-root">
      <Avatar width="6rem" />
      <section>
        <p className="name">Hello, Ashley!</p>
      </section>
      <section className="locale-container">
        <LocaleDisplay />
      </section>
      <section className="button-container">
        <button
          type="button"
          className="button-function"
          onClick={() => dispatch(openModal(MODALS.UPLOAD_CSV))}
          data-testid="button-upload"
        >
          {t('drawer.functions.uploadCSV')}
        </button>
      </section>
      <style jsx>{`
        .drawer-root {
          background: ${COLORS.BACKGROUND_DARK_PRIMARY};
          height: 100%;
          width: 100%;
          display: flex;
          flex-direction: column;
          align-items: center;
          padding: 2rem 0;
        }

        .name {
          margin-top: 2rem;
          color: ${COLORS.TEXT_DARK_PRIMARY};
        }

        .locale-container {
          padding: 0 1rem;
          margin-top: 2rem;
          text-align: center;
        }

        .button-container {
          margin-top: 2rem;
          width: 100%;
        }

        .button-function {
          width: 100%;
          background: ${COLORS.BACKGROUND_DARK_SECONDARY};
          color: ${COLORS.TEXT_DARK_PRIMARY};
          border: none;
          padding: 0.5rem 0;
        }
      `}</style>
    </aside>
  );
};

export default Drawer;
