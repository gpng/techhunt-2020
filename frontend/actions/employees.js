/* eslint-disable max-classes-per-file */
import axios from 'axios';
import { to } from '../utils';
// components
import AxioRequest from './Request';
// constants
import { API_HOST } from '../constants';

// eslint-disable-next-line import/prefer-default-export
export class PostCSV extends AxioRequest {
  async call(csvFile) {
    const uri = `${API_HOST}/users/upload`;

    const formData = new FormData();
    formData.append('file', csvFile);

    const [err, res] = await to(
      axios.post(uri, formData, {
        cancelToken: this.getCancelToken(),
        headers: {
          'content-type': 'multipart/form-data',
        },
      }),
    );
    if (!err && res.status === 200 && res.data) {
      return [null, res.data];
    }
    return [err?.response?.data?.message ?? 'server error', null];
  }
}
