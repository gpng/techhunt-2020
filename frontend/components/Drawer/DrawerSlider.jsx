import React, { useState, useRef } from 'react';
import classNames from 'classnames';
// components
import Drawer from './Drawer';
// utils
import { useOnClickOutside } from '../../utils/hooks';

const DrawerSlider = () => {
  const [open, setOpen] = useState(false);

  const containerRef = useRef(null);

  useOnClickOutside(containerRef, () => {
    if (open) {
      setOpen(false);
    }
  });

  return (
    <div className="drawer-slider-root">
      <button type="button" className="button-menu" onClick={() => setOpen(!open)}>
        hamburger
      </button>
      <div
        ref={containerRef}
        className={classNames('drawer-slider-container', {
          'drawer-slider-container--open': open,
        })}
      >
        <Drawer />
      </div>
      <style jsx>{`
        .button-drawer-slider {
        }

        .drawer-slider-container {
          position: fixed;
          height: 100%;
          width: 12rem;
          top: 0;
          left: 0;
          transform: translateX(-12rem);
          transition: all 0.2s;
        }

        .drawer-slider-container--open {
          transform: translateX(0);
        }
      `}</style>
    </div>
  );
};

export default DrawerSlider;
