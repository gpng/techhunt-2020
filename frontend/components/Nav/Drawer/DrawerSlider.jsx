import React, { useState, useRef } from 'react';
import classNames from 'classnames';
// components
import Drawer from './Drawer';
import { Menu } from '../../General/Icons';
// utils
import { useOnClickOutside } from '../../../utils/hooks';
// constants
import { COLORS } from '../../../constants/styles';

const DrawerSlider = () => {
  const [open, setOpen] = useState(false);

  const containerRef = useRef(null);

  useOnClickOutside(containerRef, () => {
    if (open) {
      setOpen(false);
    }
  });

  return (
    <div className="drawer-slider-root">
      <button
        type="button"
        className="button-menu"
        onClick={() => setOpen(!open)}
        data-testid="button-menu"
      >
        <Menu width="1rem" fill={COLORS.BACKGROUND_PRIMARY} />
      </button>
      <div
        ref={containerRef}
        className={classNames('drawer-slider-container', {
          'drawer-slider-container--open': open,
        })}
      >
        <Drawer />
      </div>
      <style jsx>{`
        .button-drawer-slider {
        }

        .button-menu {
          width: 2rem;
          height: 2rem;
          display: flex;
          justify-content: center;
          align-items: center;
          background: transparent;
          border: none;
        }

        .drawer-slider-container {
          position: fixed;
          height: 100%;
          width: 12rem;
          top: 0;
          left: 0;
          transform: translateX(-12rem);
          transition: all 0.2s;
        }

        .drawer-slider-container--open {
          transform: translateX(0);
        }
      `}</style>
    </div>
  );
};

export default DrawerSlider;
