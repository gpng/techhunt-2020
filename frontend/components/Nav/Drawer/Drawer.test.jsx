import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';
import Drawer from './Drawer';
import { renderWithContext, renderWithCustomAppContext } from '../../../utils/tests';
import { openModal } from '../../../actions/creators';
import { MODALS } from '../../../constants';

afterEach(cleanup);

it('should render', () => {
  renderWithContext(<Drawer />);
});

// test if correct classname is added
it('should match default snapshot', () => {
  const { asFragment } = renderWithContext(<Drawer />);
  expect(asFragment()).toMatchSnapshot();
});

it('should dispatch open upload action on button click', () => {
  const mockDispatch = jest.fn();
  const { getByTestId } = renderWithCustomAppContext(<Drawer />, {
    value: { dispatch: mockDispatch },
  });
  fireEvent.click(getByTestId('button-upload'));
  expect(mockDispatch).toHaveBeenCalledTimes(1);
  expect(mockDispatch).toHaveBeenCalledWith(openModal(MODALS.UPLOAD_CSV));
});
