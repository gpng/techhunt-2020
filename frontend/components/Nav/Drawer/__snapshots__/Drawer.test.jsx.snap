// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match default snapshot 1`] = `
<DocumentFragment>
  <aside
    class="jsx-3606731684 drawer-root"
  >
    <div
      class="jsx-1818524733 avatar-root"
    />
    <div
      class="jsx-3606731684 button-container"
    >
      <button
        class="jsx-3606731684 button-function"
        type="button"
      >
        Upload CSV
      </button>
    </div>
  </aside>
</DocumentFragment>
`;
